modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java

```diff
@@ -288,50 +301,134 @@ protected int processStartTag() throws Exception {
 
-		buttonTag.setCssClass("dropdown-toggle page-link");
-		buttonTag.setDisplayType("unstyled");
-		buttonTag.setLabel("...");
+				buttonTag = new ButtonTag();
 
-		buttonTag.doTag(pageContext);
+				buttonTag.setCssClass("dropdown-toggle page-link");
+				buttonTag.setDisplayType("unstyled");
+				buttonTag.setLabel("...");
 
-		jspWriter.write("</li>");
+				buttonTag.doTag(pageContext);
 
-		for (int i = ellipsisCount + 1; i < (totalPages + 1); i++) {
-			jspWriter.write("<li class=\"page-item");
+				jspWriter.write("</li>");
+			}
 
-			if (i == _activePage) {
-				jspWriter.write(" active");
+			int lastIndex = totalItems.length - 1;
+
+			int[] centerItems = Arrays.copyOfRange(
+				totalItems, Math.max(activeIndex - _ellipsisBuffer, 1),
+				Math.min(activeIndex + _ellipsisBuffer + 1, lastIndex));
+
+			for (int itemToRender : centerItems) {
+				jspWriter.write("<li class=\"page-item");
+
+				if (itemToRender == _activePage) {
+					jspWriter.write(" active");
+				}
+
+				if ((_disabledPages != null) &&
+					_disabledPages.contains(Integer.valueOf(itemToRender))) {
+
+					jspWriter.write(" disabled");
+				}
+
+				jspWriter.write("\">");
+
+				buttonTag = new ButtonTag();
+
+				buttonTag.setCssClass("page-link");
+				buttonTag.setDisplayType("unstyled");
+				buttonTag.setLabel(String.valueOf(itemToRender));
+
+				buttonTag.doTag(pageContext);
 			}
 
-			if ((_disabledPages != null) &&
-				_disabledPages.contains(Integer.valueOf(i))) {
+			int rightIndex = activeIndex + _ellipsisBuffer + 1;
 
-				jspWriter.write(" disabled");
+			if (rightIndex > lastIndex) {
+				rightIndex = lastIndex;
 			}
 
-			jspWriter.write("\">");
+			int[] rightItems = Arrays.copyOfRange(
+				totalItems, rightIndex, Math.max(lastIndex, rightIndex));
 
-			buttonTag = new ButtonTag();
+			if (rightItems.length > 1) {
+				jspWriter.write("<li class=\"dropdown page-item\">");
 
-			buttonTag.setCssClass("page-link");
-			buttonTag.setDisplayType("unstyled");
+				buttonTag = new ButtonTag();
 
-			buttonTag.setLabel(String.valueOf(i));
+				buttonTag.setCssClass("dropdown-toggle page-link");
+				buttonTag.setDisplayType("unstyled");
+				buttonTag.setLabel("...");
 
-			buttonTag.doTag(pageContext);
+				buttonTag.doTag(pageContext);
+
+				jspWriter.write("</li>");
+			}
+
+			if (totalPages > 1) {
+				int lastItem = totalItems[lastIndex];
+
+				jspWriter.write("<li class=\"page-item");
+
+				if (lastItem == _activePage) {
+					jspWriter.write(" active");
+				}
+
+				if ((_disabledPages != null) &&
+					_disabledPages.contains(Integer.valueOf(lastItem))) {
+
+					jspWriter.write(" disabled");
+				}
+
+				jspWriter.write("\">");
+
+				buttonTag = new ButtonTag();
 
-			jspWriter.write("</li>");
+				buttonTag.setCssClass("page-link");
+				buttonTag.setDisplayType("unstyled");
+				buttonTag.setLabel(String.valueOf(lastItem));
+
+				buttonTag.doTag(pageContext);
+			}
+		}
+		else {
+			for (int i = 1; i < (totalPages + 1); i++) {
```

Comment:
I think you can remove the parenthesis, although maybe is the SourceFormatter who is adding them...

Reply:
I'm not sure, but I think it was an SF thing